gem/optimise.py

```diff
@@ -493,3 +592,105 @@ def aggressive_unroll(expression):
+@count_flop_node.register(ListTensor)
+@count_flop_node.register(FlexiblyIndexed)
+@count_flop_node.register(LogicalNot)
+@count_flop_node.register(LogicalAnd)
+@count_flop_node.register(LogicalOr)
+@count_flop_node.register(Conditional)
+def count_flop_node_zero(node):
+    return 0
+
+
+@count_flop_node.register(Power)
+@count_flop_node.register(Comparison)
+@count_flop_node.register(Sum)
+@count_flop_node.register(Product)
+@count_flop_node.register(Division)
+@count_flop_node.register(MathFunction)
+def count_flop_node_single(node):
+    return numpy.prod([idx.extent for idx in node.free_indices])
+
+
+@count_flop_node.register(IndexSum)
+def count_flop_node_index_sum(node):
+    return numpy.prod([idx.extent for idx in node.multiindex + node.free_indices])
+
+
+def count_flop(node):
+    """
+    Count the total floating point operations required to compute a gem node.
+    This function assumes that all subnodes that occur more than once induce a
+    temporary, and are therefore only computed once.
+    """
+    # TODO: Add tests for this function (number still somewhat diffferent from COFFEE visitor)
+    return sum(map(count_flop_node, traversal([node])))
+
+
+def fast_sum_factorise(sum_indices, factors):
+    return sum_factorise(*delta_elimination(list(reversed(sum_indices)), factors))
+
+
+def optimise(node, quadrature_multiindex, argument_multiindices):
+    from gem.refactorise import ATOMIC, COMPOUND, OTHER, collect_monomials
+    argument_indices = tuple([i for indices in argument_multiindices for i in indices])
+
+    def classify(argument_indices, expression):
+        if isinstance(expression, Conditional):
+            return ATOMIC
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, Indexed):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+    classifier = partial(classify, set(argument_indices))
+
+    monomial_sum, = collect_monomials([node], classifier)
+    monomial_sum.argument_indices = argument_indices
+
+    optimal_atomics = []  # [(sum_indices, optimal_atomics))]
+    other_atomics = []  # [(sum_indices, other_atomics))]
+    for sum_indices in monomial_sum.all_sum_indices():
+        atomics = monomial_sum.find_optimal_atomics(sum_indices)
+        optimal_atomics.extend([(sum_indices, _atomic) for _atomic in atomics[0]])
+        other_atomics.extend([(sum_indices, _atomic) for _atomic in atomics[1]])
+    optimal_atomics.extend(other_atomics)  # sequence of atomics to factorise
+    # This algorithm is O(2^N), where N = len(optimal_atomics)
+    # we could truncate the optimal_atomics list at say 10
+    monomial_sum.factorise_atomics(optimal_atomics)
+
+    return monomial_sum.to_expression()
+
+
+def optimise_expressions(expressions, quadrature_indices, argument_indices):
```

Comment:
I sort of feel like the idea behind the modes was that the mode would contain these kind of "mode-specific" optimisation passes.  But maybe I misthought.